dev/publish_changelog.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/python
+#
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Provides facilities for publishing changelogs to
+# our github.io site.
+
+import argparse
+import datetime
+import os
+import re
+import sys
+
+from spinnaker.run import check_run_quick
+
+# Path to the posts directory in the spinnaker.github.io site.
+POSTS_DIR = '_posts'
+
+class ChangelogPublisher():
+
+  def __init__(self, options):
+    self.__version = options.version
+    self.__githubio_repo_uri = options.githubio_repo_uri
+    self.__repo_name = ''
+    self.__changelog_gist_uri = options.changelog_gist_uri
+
+
+  def __checkout_githubio_repo(self):
+    """Clones the spinnaker.github.io git repo.
+    """
+    check_run_quick('git clone {0}'.format(self.__githubio_repo_uri))
+    self.__repo_name = os.path.basename(self.__githubio_repo_uri)
+    if self.__repo_name.endswith('.git'):
+      self.__repo_name = re.sub('.git$', '', self.__repo_name)
+
+  def __format_changelog_post(self):
+    # Initialized with 'front matter' necessary for the post.
+    timestamp = '{:%Y-%m-%d %H:%M:%S}'.format(datetime.datetime.now())
+    post_lines = [
+      '---',
+      'title: Spinnaker Changelog {version}'.format(version=self.__version),
+      'date: {date}'.format(date=timestamp),
+      'categories: changelogs',
+      '---',
+      ''
+    ]
+
+    post_lines.append('<script src="{uri}"></script>'.format(uri=self.__changelog_gist_uri))
+    post = '\n'.join(post_lines)
+    return post
+
+  def __publish_post(self, post_content):
+    day = '{:%Y-%m-%d}'.format(datetime.datetime.now())
+    post_name = '{day}-{version}-changelog.md'.format(day=day, version=self.__version)
+    post_path = os.path.join(self.__repo_name, POSTS_DIR, post_name)
+    # Path to post file relative to the git root.
+    post_rel_path = os.path.join(POSTS_DIR, post_name)
+    with open(post_path, 'w') as post_file:
+      post_file.write(post_content)
+
+    check_run_quick('git -C {0} add {1}'.format(self.__repo_name, post_rel_path))
+    message = 'Changelog for version {0}'.format(self.__version)
```

Comment:
I'd add a comment that this commit was auto-generated by a script (you can get the name of this current script by using `__file__`, a global constant)